internal/supervisor/scheduler.go

```diff
@@ -0,0 +1,78 @@
+package supervisor
+
+import (
+	"time"
+)
+
+// Scheduler schedules a doer for periodic, serial execution.
+type Scheduler struct {
+	doer   Doer
+	period time.Duration
+	notify chan bool
+	stop   chan bool
+	done   chan bool
+}
+
+// NewScheduler creates a new scheduler.
+//
+// If the period is zero, the scheduler is never run, even if manually run. Otherwise, the period
+// specifies how long to wait after its last successful execution.
+func NewScheduler(doer Doer, period time.Duration) *Scheduler {
+	s := &Scheduler{
+		doer:   doer,
+		period: period,
+		notify: make(chan bool, 1),
+		stop:   make(chan bool),
+		done:   make(chan bool),
+	}
+
+	go s.run()
+
+	return s
+}
+
+// Do requests an execution of the scheduled doer.
+//
+// The scheduler will never be run if the period is 0. If already running, the doer will be run
+// again when done. Multiple calls to Notify while the doer is running will only trigger a single
+// additional execution. Do never blocks.
+func (s *Scheduler) Do() error {
+	select {
+	case s.notify <- true:
+	default:
+	}
+
+	return nil
+}
+
+// run is the main thread of the scheduler and responsible for triggering the doer as required.
+func (s *Scheduler) run() {
+	for {
+		// Enabling doing only if a non-zero interval was configured.
+		var poll <-chan time.Time
+		var notify <-chan bool
+		if s.period > 0 {
+			poll = time.After(s.period)
+			notify = s.notify
```

Comment:
Shouldn't notify be assigned anyway? So that manual Do functions if period is 0? Otherwise what's the point of a period of 0?

Reply:
That was the original behaviour, but it proved to be somewhat useless to run the scheduler manually -- it would only happen on an API call, and doesn't "guarantee" that changes are made (e.g. if locks are acquired).

My primary use case here is to be able to run an instance against a shared database /without/ triggering background jobs unintentionally (e.g. by effectively adding a node to the cluster). Towards that end, setting a period of `0` disables the scheduler altogether, and is currently documented as such in the constructor function:
```
// If the period is zero, the scheduler is never run, even if manually run. Otherwise, the period
// specifies how long to wait after its last successful execution.
```

Maybe the manual `notify` is overkill, and I should just set a lower period so that the need to poke becomes irrelevant?

Reply:
Ok. That makes sense. I seen that it was documented but I thought it was strange enough to ask about.

Probably overkill, but I do like the concept of my commands poking the scheduler no matter the period.